### PR TITLE
xrt ops use bfc_allocator instead

### PR DESCRIPTION
**Pull request by yaochengji**
XRT ops use `cuMemAlloc` directly, which will hurt the performance a lot.

This PR changes the cuda allocator to bfc allocator.

For Resnet-50 fp16 training with batchsize 256 in Tesla V100 GPU, the performance could increase from `730 images/s` to `1290 images/s`. Please refer to the [torch/xla issue](https://github.com/pytorch/xla/issues/2614) for more details.

### CODE REVIEW

**Comment by yaochengji** [comment]
@googlebot I signed it

**Comment by byronyi** [comment]
Ping @dlibenzi; wonder if you are interested in reviewing this PR.

**Comment by yaochengji** [comment]
Hi @gbaned @joker-eph,

all the xrt unit tests have passed, could you help review?

**Comment by yaochengji** [comment]
@yunxing could you help review?

**Comment by yaochengji** [comment]
@gbaned more than 2 weeks have passed but no one help review. Could you help find any other reviewers?

**Comment by mihaimaruseac** [comment]
`tensorflow/compiler` code should only be reviewed by people from the compilers teams.

**Comment by tatianashp** [comment]
@sanjoy Can you please take a look at this PR or find an appropriate reviewer.

File: tensorflow/core/common_runtime/gpu/gpu_process_state.h [hunk 1]
```diff
@@ -89,6 +89,10 @@ class GPUProcessState {
     return gpu_allocators_.size();
   }
 
+  Allocator* GetConstructedGPUAllocator(int ordinal) {
```
Comment (JackCaoG) [inline tensorflow/core/common_runtime/gpu/gpu_process_state.h:92]: Why not just use the existing `GPUProcessState::GetGPUAllocator`? This seems like a hack.

Comment (yaochengji) [reply tensorflow/core/common_runtime/gpu/gpu_process_state.h]: Oh, I thought `GPUProcessState::GetGPUAllocator` was only used to create a new allocator. And it is fixed now.

**Comment by JackCaoG** [comment]
There are some compilations for this pr to merge externally (if we decide to take it). Helper functions like `GetInputTupleAllocations` is being used by the internal `XRT_TPU_ops` which is is invisible from the outside. I recently open sourced the [tpu_compile_op](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xrt/kernels/tpu_compile_ops.cc) and [tpu_execute_op](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xrt/kernels/tpu_execute_op.cc) which also uses these helper functions. However the internal `XRT_TPU_ops` is still being used.

I would suggest you wait for a couple days before I open sourced all of the `tpu_ops` and update those files. It is likely I will take sometime to review this pr since I am not as familiar with the gpu stuff. If by that time we still find this pr to be valueable, I might have to take this diff and update the internal usage of helper functions and submit it internally.

**Comment by yaochengji** [comment]
Thanks for following up with this pr. And we're experimentally using torch/xla to accelerate pytorch training on GPU with this patch in the company. Currently we maintain our own branch of torch/xla and tensorflow, and surely we could wait for your side.

**Comment by tensorflowbutler** [comment]
It has been 24 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by JackCaoG** [comment]
A bit busy recently, will try to circle back to this pr

**Comment by gbaned** [comment]
@JackCaoG, @yaochengji  Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by JackCaoG** [comment]
Sorry I missed the last email. @yaochengji There has been couple more TPU related ops added to `tensorflow/compiler/xrt/kernels/`. If you could update the signatures  for those ops, I can take this diff and update the internal ops and merge it internally.

**Comment by yaochengji** [comment]
> Sorry I missed the last email. @yaochengji There has been couple more TPU related ops added to `tensorflow/compiler/xrt/kernels/`. If you could update the signatures for those ops, I can take this diff and update the internal ops and merge it internally.

Thanks @JackCaoG. I will update it in the following days.

**Comment by yaochengji** [comment]
@JackCaoG, I only updated the [branch](https://github.com/yaochengji/tensorflow/tree/fix-xrt) and fixed some compilation issues. Xrt unit test could pass. It seems that there's no need to update the GPU related ops' signatures

**Comment by JackCaoG** [comment]
Thanks, I think my you meant `TPU related op`, let me circle back and check whether those ops uses the helper function you modifyied.

**Comment by yaochengji** [comment]
@JackCaoG Hello, Any progress about the commit? Or should I submit a new PR?

**Comment by codeislife99** [comment]
@yaochengji Does the XRT allocator allocate memory again and again in each fwd/bwd pass vs the BFC allocator that does it once and reuse it again? If what I said doesn't sound correct do you mind elaborating why `cuMemAlloc` is suboptimal ?

**Comment by yaochengji** [comment]
@codeadmin-peritiae  on the high level, both the XRT original allocator and BFC allocator will allocate memory again and again, but their implementation differs. The original one allocates and releases money by cuda driver API, which sync all cuda streams and hurt the performance. And the BFC allocator only calls the cuda driver api and allocate a large chunk of cuda memory when initializing. Every iteration, it could get part of it and doesn't need to call the driver api.

**Comment by codeislife99** [comment]
@yaochengji  Oh I see thanks so much for your explanation and also for this PR. I am assuming you must have made many benchmarks across different models with this fix on your custom versions of PT-XLA. I see you shared the performance improvement for Resnet-50. Do you mind sharing other benchmarks as well that you have done(maybe across different batch sizes, detection models etc. or anything else) It would be great to see those numbers compared with Vanilla PT.

**Comment by yaochengji** [comment]
@codeislife99

**Comment by codeislife99** [comment]
What's the batch size for those numbers and what's the CUDA architecture ( V100 / T4 etc.) And number of GPUs.

**Comment by yaochengji** [comment]
@codeislife99

File: tensorflow/compiler/xrt/kernels/xrt_execute_op.cc [hunk 2]
```diff
@@ -388,24 +391,24 @@ xla::StatusOr<RefPtr<XRTTupleAllocation>> ExecuteComputation(
   // memory, until either the runfn can run, or we run out of freeable memory.
   return memory_manager->Run<RefPtr<XRTTupleAllocation>>(
       runfn, device_ref->backend(), device_ref->device_ordinal(),
-      /*requested_free_size=*/0);
+      /*requested_free_size=*/0, allocator);
```
Comment (JackCaoG) [inline tensorflow/compiler/xrt/kernels/xrt_execute_op.cc:394]: Instead of passing `allocator` all the way down here, couldn't we call `device_ref->GetMemoryAllocator()` directly?

File: tensorflow/compiler/xrt/kernels/xrt_execute_op.cc [hunk 3]
```diff
@@ -567,11 +571,13 @@ Status XRTExecuteChainedOp::DoWork(OpKernelContext* context) {
 
     return ExecuteComputation(
         context, memory_manager.get(), &device_ref, executable, input_tuples,
-        /*release_inputs=*/false, stream, rng_seed, config.common_config());
+        /*release_inputs=*/false, stream, rng_seed, config.common_config(),
+        device_ref.GetMemoryAllocator());
```
Comment (JackCaoG) [inline tensorflow/compiler/xrt/kernels/xrt_execute_op.cc:575]: ditto, we don't really need to pass `allocator` here since we also pass `device_ref`

File: tensorflow/compiler/xrt/xrt_device.h [hunk 4]
```diff
@@ -50,19 +52,32 @@ class XRTGenericDeviceAccessor {
     xla::LocalClient* client() const { return client_; }
     xla::Backend* backend() { return client_->mutable_backend(); }
     int device_ordinal() const { return ordinal_; }
+    // XRTCompileOp needs to use a memory allocator constructed with stream.
+    se::DeviceMemoryAllocator* GetMemoryAllocator(OpKernelContext* ctx);
```
Comment (JackCaoG) [inline tensorflow/compiler/xrt/xrt_device.h:56]: hmm, seems like `XRTTpuDeviceAccessor` should also implement `GetMemoryAllocator` since stateOP implementation is shared between tpu and cpu/gpu. I think it OK to always return `client_->mutable_backend()->memory_allocator()`

File: tensorflow/compiler/xrt/xrt_device.h [hunk 5]
```diff
@@ -50,19 +52,32 @@ class XRTGenericDeviceAccessor {
...
+                 const std::string& platform_name) {
       client_ = client;
       ordinal_ = ordinal;
+      platform_name_ = platform_name;
     }
 
     xla::LocalClient* client_ = nullptr;
     int ordinal_ = 0;
+    std::string platform_name_;
+    se::Platform::Id platform_id_ = nullptr;
```
Comment (JackCaoG) [inline tensorflow/compiler/xrt/xrt_device.h:75]: doesn't seem like we need `platform_id_`

File: tensorflow/compiler/xrt/xrt_memory_manager.h [hunk 6]
```diff
@@ -168,7 +173,8 @@ xla::StatusOr<T> XRTMemoryManager::Run(
     if (result_or.status().code() != error::RESOURCE_EXHAUSTED) {
       return result_or;
     }
-    TF_RETURN_IF_ERROR(TryFreeMemoryStep(&mrctx, result_or.status()));
+    TF_RETURN_IF_ERROR(
+        TryFreeMemoryStep(&mrctx, result_or.status(), allocator));
```
Comment (JackCaoG) [inline tensorflow/compiler/xrt/xrt_memory_manager.h:177]: might be better to add `allocator` to `MemoryReclaimContext`, wdyt?

File: tensorflow/compiler/xrt/xrt_state.cc [hunk 7]
```diff
@@ -184,45 +185,43 @@ void XRTTupleAllocation::ReleaseBuffers() {
...
-      ->InitializeFromShapedBuffer(shaped_buffer, backend->memory_allocator(),
-                                   device_ordinal);
+      ->InitializeFromShapedBuffer(shaped_buffer, allocator, device_ordinal);
   (*allocation)->SetDeviceMemorySize();
   return Status::OK();
 }
 
 /*static*/ Status XRTTupleAllocation::CreateFromBuffer(
     const xla::ShapedBuffer& shaped_buffer, const xla::Shape& on_host_shape,
     const xla::Shape& on_device_shape, xla::Backend* backend,
```
Comment (JackCaoG) [inline tensorflow/compiler/xrt/xrt_state.cc:222]: doesn't seems like we need to pass`backend` anymore

**Review by JackCaoG — COMMENTED** [review]
@gbaned Could you reopen this pr? Thanks @yaochengji! Could you do a TPUVM build with `TPUVM_MODE=1 python setup.py install` for pt/xla? I think this change will break tpu version of xrt ops.

File: tensorflow/compiler/xrt/xrt_device.cc [hunk 8]
```diff
@@ -76,16 +76,65 @@ XRTGenericDeviceAccessor::GetOrCreateCompilationCache(
...
+  scoped_ref->Acquire(metadata->client(), metadata->device_ordinal(),
+                      metadata->platform()->Name());
   return Status::OK();
 }
 
+mutex XRTGenericDeviceAccessor::ScopedRef::mutex_;
+std::map<void*, std::unique_ptr<se::TfAllocatorAdapter>>
+    XRTGenericDeviceAccessor::ScopedRef::compile_cuda_allocators_;
+std::map<std::string, std::unique_ptr<se::TfAllocatorAdapter>>
+    XRTGenericDeviceAccessor::ScopedRef::other_cuda_allocators_;
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply tensorflow/compiler/xrt/xrt_device.cc:97]: Could you do a TPUVM build with TPUVM_MODE=1 python setup.py install for pt/xla? I think this change will break tpu version of xrt ops.

File: tensorflow/compiler/xrt/xrt_device.cc [hunk 9]
```diff
@@ -76,16 +76,65 @@ XRTGenericDeviceAccessor::GetOrCreateCompilationCache(
...
+
+se::DeviceMemoryAllocator*
+XRTGenericDeviceAccessor::ScopedRef::GetMemoryAllocator() {
+  if (platform_name_ != "CUDA") {
+    return client_->mutable_backend()->memory_allocator();
+  }
+  if (!other_cuda_allocators_.count(platform_name_)) {
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply tensorflow/compiler/xrt/xrt_device.cc:104]: 1. since the optimization is mainly for CUDA, cpu and TPU could use the original allocator

File: tensorflow/compiler/xrt/xrt_device.h [hunk 10]
```diff
@@ -50,19 +52,32 @@ class XRTGenericDeviceAccessor {
...
+    static mutex mutex_;
+    static std::map<void*, std::unique_ptr<se::TfAllocatorAdapter>>
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply tensorflow/compiler/xrt/xrt_device.h:77]: Done.

File: tensorflow/compiler/xrt/xrt_memory_manager.cc [hunk 11]
```diff
@@ -265,8 +268,8 @@ void XRTMemoryManager::ReleaseAllAllocations() {
 }
 
 xla::StatusOr<se::OwningDeviceMemory> XRTMemoryManager::Allocate(
-    xla::Backend* backend, int device_ordinal, size_t size) {
-  se::DeviceMemoryAllocator* allocator = backend->memory_allocator();
+    xla::Backend* backend, int device_ordinal, size_t size,
+    se::DeviceMemoryAllocator* allocator) {
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply tensorflow/compiler/xrt/xrt_memory_manager.cc:272]: yeah, we don't need to pass it.

File: tensorflow/compiler/xrt/kernels/xrt_execute_op.cc (hunk 2 above)
Comment (yaochengji) [reply tensorflow/compiler/xrt/kernels/xrt_execute_op.cc:394]: Done.

File: tensorflow/compiler/xrt/kernels/xrt_execute_op.cc (hunk 3 above)
Comment (yaochengji) [reply tensorflow/compiler/xrt/kernels/xrt_execute_op.cc:575]: Done.

File: tensorflow/compiler/xrt/xrt_device.h (hunk 4 above)
Comment (yaochengji) [reply tensorflow/compiler/xrt/xrt_device.h:56]: Done.

File: tensorflow/compiler/xrt/xrt_device.h (hunk 5 above)
Comment (yaochengji) [reply tensorflow/compiler/xrt/xrt_device.h:75]: Done.

File: tensorflow/compiler/xrt/xrt_memory_manager.h (hunk 6 above)
Comment (yaochengji) [reply tensorflow/compiler/xrt/xrt_memory_manager.h:177]: Because `ScopedRef` is more accessible from everywhere. But putting allocator to `MemoryReclaimContext` seems more reasonable.

File: tensorflow/compiler/xrt/xrt_state.cc (hunk 7 above)
Comment (yaochengji) [reply tensorflow/compiler/xrt/xrt_state.cc:222]: Now TPU side needs this.

**Comment by yaochengji** [comment]
@JackCaoG added a new commit to fix the TPU compatible issue and most of the comments.

**Comment by JackCaoG** [comment]
@yaochengji Thanks! Will take a look today or tmr.

**Comment by JackCaoG** [comment]
@yaochengji I think you forgot to update  the pr, the change I am seeing is out of date.

**Comment by yaochengji** [comment]
Oh, because the PR is closed, it will not update automatically.

See [my branch](https://github.com/yaochengji/tensorflow/tree/fix-xrt), or should I open a new one?

**Comment by JackCaoG** [comment]
@yaochengji Sounds good, could you open a new PR?

**Comment by yaochengji** [comment]
> @yaochengji Sounds good, could you open a new PR?

reopened, https://github.com/tensorflow/tensorflow/pull/48832
